Unfortunately, there isn't much useful information from the CI logs (or I didn't find it).

Thus, I ran the privileged tests on my dev machine 100+ times without hitting a failure (`for i in $(seq 100); do sudo make tests-privileged || exit 1; done`), arch linux with kernel `5.4.15-arch1-1`.

Then I tried running the privileged tests on a dev VM (Ubuntu with kernel `4.9.160-0409160-generic`) and there I managed to get a failure and a trace:
<details>

```
vagrant@runtime1:~/go/src/github.com/cilium/cilium$ sudo make tests-privileged
# cilium-map-migrate is a dependency of some unit tests.
make -C bpf cilium-map-migrate
make[1]: Entering directory '/home/vagrant/go/src/github.com/cilium/cilium/bpf'
make[1]: 'cilium-map-migrate' is up to date.
make[1]: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium/bpf'
go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/cilium/cmd -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/bpf -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/config -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/ipsec -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/probes -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/route -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/loader -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/fqdn/dnsproxy -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/maps/ctmap -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/maps/eppolicymap -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/maps/policymap -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/mtu -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;
=== RUN   Test
START: bpf_ipcache_get_test.go:58: BPFIPCacheGetSuite.TestGetLPMValue
PASS: bpf_ipcache_get_test.go:58: BPFIPCacheGetSuite.TestGetLPMValue	0.000s

START: bpf_ipcache_get_test.go:32: BPFIPCacheGetSuite.TestGetPrefix
PASS: bpf_ipcache_get_test.go:32: BPFIPCacheGetSuite.TestGetPrefix	0.002s

START: helpers_test.go:39: CMDHelpersSuite.TestExpandNestedJSON
PASS: helpers_test.go:39: CMDHelpersSuite.TestExpandNestedJSON	0.001s

START: helpers_test.go:576: CMDHelpersSuite.TestParsePolicyUpdateArgsHelper
PASS: helpers_test.go:576: CMDHelpersSuite.TestParsePolicyUpdateArgsHelper	0.000s

START: helpers_test.go:551: CMDHelpersSuite.TestParseTrafficString
PASS: helpers_test.go:551: CMDHelpersSuite.TestParseTrafficString	0.000s

OK: 5 passed
--- PASS: Test (0.00s)
PASS
ok  	github.com/cilium/cilium/cilium/cmd	0.033s
level=info msg="Detected mounted BPF filesystem at /sys/fs/bpf" subsys=bpf
=== RUN   Test
START: map_linux_test.go:250: BPFPrivilegedTestSuite.TestBasicManipulation
PASS: map_linux_test.go:250: BPFPrivilegedTestSuite.TestBasicManipulation	0.000s

START: map_linux_test.go:521: BPFPrivilegedTestSuite.TestCheckAndUpgrade
level=warning msg="Flags mismatch for BPF map" file-path=/sys/fs/bpf/tc/globals/cilium_test_upgrade new=0 old=1 subsys=bpf
level=warning msg="Removing map to allow for property upgrade (expect map data loss)" file-path=/sys/fs/bpf/tc/globals/cilium_test_upgrade subsys=bpf
PASS: map_linux_test.go:521: BPFPrivilegedTestSuite.TestCheckAndUpgrade	0.001s

START: map_linux_test.go:481: BPFPrivilegedTestSuite.TestDeleteAll
PASS: map_linux_test.go:481: BPFPrivilegedTestSuite.TestDeleteAll	0.000s

START: map_linux_test.go:319: BPFPrivilegedTestSuite.TestDump
PASS: map_linux_test.go:319: BPFPrivilegedTestSuite.TestDump	0.000s

START: map_linux_test.go:390: BPFPrivilegedTestSuite.TestDumpReliablyWithCallback
panic: test timed out after 6m0s

goroutine 57 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1377 +0xdf
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x44

goroutine 1 [chan receive, 6 minutes]:
testing.(*T).Run(0xc0002fe100, 0xd18416, 0x4, 0xd4e2e8, 0x4c44f6)
	/usr/local/go/src/testing/testing.go:961 +0x377
testing.runTests.func1(0xc0002fe000)
	/usr/local/go/src/testing/testing.go:1202 +0x78
testing.tRunner(0xc0002fe000, 0xc000621ca0)
	/usr/local/go/src/testing/testing.go:909 +0xc9
testing.runTests(0xc0003f6380, 0x139b6b0, 0x1, 0x1, 0x0)
	/usr/local/go/src/testing/testing.go:1200 +0x2a7
testing.(*M).Run(0xc0004ea900, 0x0)
	/usr/local/go/src/testing/testing.go:1117 +0x176
github.com/cilium/cilium/pkg/bpf.runTests(0xc0004ea900, 0x0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:88 +0x126
github.com/cilium/cilium/pkg/bpf.TestMain(0xc0004ea900)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:92 +0x2f
main.main()
	_testmain.go:42 +0x135

goroutine 83 [chan receive, 6 minutes]:
gopkg.in/check%2ev1.(*suiteRunner).runTest(...)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:819
gopkg.in/check%2ev1.(*suiteRunner).run(0xc000436080, 0x13d0ce0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:624 +0x11d
gopkg.in/check%2ev1.Run(0xcbf880, 0x13d0ce0, 0xc00007c708, 0xd0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:92 +0x4d
gopkg.in/check%2ev1.RunAll(0xc00008ef08, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:84 +0x96
gopkg.in/check%2ev1.TestingT(0xc0002fe100)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:72 +0x395
github.com/cilium/cilium/pkg/bpf.Test(0xc0002fe100)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:34 +0x2b
testing.tRunner(0xc0002fe100, 0xd4e2e8)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350

goroutine 84 [select, 6 minutes]:
gopkg.in/check%2ev1.(*resultTracker)._loopRoutine(0xc0002c2360)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:470 +0xc4
created by gopkg.in/check%2ev1.(*resultTracker).start
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:450 +0x3f

goroutine 89 [semacquire, 6 minutes]:
sync.runtime_Semacquire(0xc000349408)
	/usr/local/go/src/runtime/sema.go:56 +0x42
sync.(*WaitGroup).Wait(0xc000349400)
	/usr/local/go/src/sync/waitgroup.go:130 +0x64
github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback(0x13d0ce0, 0xc000308780)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:478 +0x478
reflect.Value.call(0xcbf880, 0x13d0ce0, 0x1213, 0xd182fa, 0x4, 0xc00008df30, 0x1, 0x1, 0x13d2be0, 0xc00008ddd0, ...)
	/usr/local/go/src/reflect/value.go:460 +0x5f6
reflect.Value.Call(0xcbf880, 0x13d0ce0, 0x1213, 0xc00008df30, 0x1, 0x1, 0x0, 0xe, 0x8)
	/usr/local/go/src/reflect/value.go:321 +0xb4
gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1(0xc000308780)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0x5e4
gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1(0xc000436080, 0xc000308780, 0xc0003f72a0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0xa2
created by gopkg.in/check%2ev1.(*suiteRunner).forkCall
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:672 +0x1fb

goroutine 90 [runnable]:
github.com/cilium/cilium/pkg/bpf.(*Map).Update.func1(0xc000308870, 0xc000093ee8, 0xe44940, 0xc0006045ac, 0xe3ff80, 0xc0006045b0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:810 +0x132
github.com/cilium/cilium/pkg/bpf.(*Map).Update(0xc000308870, 0xe44940, 0xc0006045ac, 0xe3ff80, 0xc0006045b0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:826 +0x333
github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback.func2(0xc000349400, 0xc0002fa540, 0xc0002fa5a0, 0xc000308870, 0xc000308780)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:432 +0x1ff
created by github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:422 +0x3df

goroutine 72 [syscall, 5 minutes]:
syscall.Syscall(0x141, 0x4, 0xc000428d68, 0x20, 0xc116a0, 0xc0004d6180, 0xc0008798b0)
	/usr/local/go/src/syscall/asm_linux_amd64.s:18 +0x5
github.com/cilium/cilium/pkg/bpf.GetNextKeyFromPointers(0x5, 0xc000428d68, 0x20, 0xc00088e6f0, 0x4)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/bpf_linux.go:218 +0x7a
github.com/cilium/cilium/pkg/bpf.(*Map).DumpReliablyWithCallback(0xc000308870, 0xc000428f40, 0xc000428f58, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:697 +0x439
github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback.func3(0xc000349400, 0xc000000100, 0xc000308780, 0xc000308870, 0xc0002fa5a0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:468 +0x315
created by github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:447 +0x467
FAIL	github.com/cilium/cilium/pkg/bpf	360.125s
FAIL
Makefile:97: recipe for target 'tests-privileged' failed
make: *** [tests-privileged] Error 1
```
</details>

`goroutine 90` and `goroutine 72` are the interesting ones here. We can see that `goroutine 90` was in the process of updating the map (`github.com/cilium/cilium/pkg/bpf.(*Map).Update()`) while `goroutine 72` was doing the `BPF_MAP_GET_NEXT_KEY` syscall in `github.com/cilium/cilium/pkg/bpf.GetNextKeyFromPointers()`.

Then another attempt, stuck again:
<details>

```
vagrant@runtime1:~/go/src/github.com/cilium/cilium$ sudo make tests-privileged
# cilium-map-migrate is a dependency of some unit tests.
make -C bpf cilium-map-migrate
make[1]: Entering directory '/home/vagrant/go/src/github.com/cilium/cilium/bpf'
make[1]: 'cilium-map-migrate' is up to date.
make[1]: Leaving directory '/home/vagrant/go/src/github.com/cilium/cilium/bpf'
go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/cilium/cmd -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/bpf -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/config -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/ipsec -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/probes -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/linux/route -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/datapath/loader -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/fqdn/dnsproxy -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/maps/ctmap -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/maps/eppolicymap -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/maps/policymap -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;  go test -mod=vendor -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/cilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/vagrant/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" github.com/cilium/cilium/pkg/mtu -test.v -timeout 360s -check.vv -tags=privileged_tests || exit 1;
=== RUN   Test
START: bpf_ipcache_get_test.go:58: BPFIPCacheGetSuite.TestGetLPMValue
PASS: bpf_ipcache_get_test.go:58: BPFIPCacheGetSuite.TestGetLPMValue	0.000s

START: bpf_ipcache_get_test.go:32: BPFIPCacheGetSuite.TestGetPrefix
PASS: bpf_ipcache_get_test.go:32: BPFIPCacheGetSuite.TestGetPrefix	0.001s

START: helpers_test.go:39: CMDHelpersSuite.TestExpandNestedJSON
PASS: helpers_test.go:39: CMDHelpersSuite.TestExpandNestedJSON	0.001s

START: helpers_test.go:576: CMDHelpersSuite.TestParsePolicyUpdateArgsHelper
PASS: helpers_test.go:576: CMDHelpersSuite.TestParsePolicyUpdateArgsHelper	0.000s

START: helpers_test.go:551: CMDHelpersSuite.TestParseTrafficString
PASS: helpers_test.go:551: CMDHelpersSuite.TestParseTrafficString	0.000s

OK: 5 passed
--- PASS: Test (0.00s)
PASS
ok  	github.com/cilium/cilium/cilium/cmd	0.038s
level=info msg="Detected mounted BPF filesystem at /sys/fs/bpf" subsys=bpf
=== RUN   Test
START: map_linux_test.go:250: BPFPrivilegedTestSuite.TestBasicManipulation
PASS: map_linux_test.go:250: BPFPrivilegedTestSuite.TestBasicManipulation	0.000s

START: map_linux_test.go:521: BPFPrivilegedTestSuite.TestCheckAndUpgrade
level=warning msg="Flags mismatch for BPF map" file-path=/sys/fs/bpf/tc/globals/cilium_test_upgrade new=0 old=1 subsys=bpf
level=warning msg="Removing map to allow for property upgrade (expect map data loss)" file-path=/sys/fs/bpf/tc/globals/cilium_test_upgrade subsys=bpf
PASS: map_linux_test.go:521: BPFPrivilegedTestSuite.TestCheckAndUpgrade	0.000s

START: map_linux_test.go:481: BPFPrivilegedTestSuite.TestDeleteAll
PASS: map_linux_test.go:481: BPFPrivilegedTestSuite.TestDeleteAll	0.000s

START: map_linux_test.go:319: BPFPrivilegedTestSuite.TestDump
PASS: map_linux_test.go:319: BPFPrivilegedTestSuite.TestDump	0.000s

START: map_linux_test.go:390: BPFPrivilegedTestSuite.TestDumpReliablyWithCallback
panic: test timed out after 6m0s

goroutine 38 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1377 +0xdf
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x44

goroutine 1 [chan receive, 6 minutes]:
testing.(*T).Run(0xc000102600, 0xd18416, 0x4, 0xd4e2e8, 0x4c44f6)
	/usr/local/go/src/testing/testing.go:961 +0x377
testing.runTests.func1(0xc000102400)
	/usr/local/go/src/testing/testing.go:1202 +0x78
testing.tRunner(0xc000102400, 0xc000609ca0)
	/usr/local/go/src/testing/testing.go:909 +0xc9
testing.runTests(0xc0000d9ec0, 0x139b6b0, 0x1, 0x1, 0x0)
	/usr/local/go/src/testing/testing.go:1200 +0x2a7
testing.(*M).Run(0xc0004e1f00, 0x0)
	/usr/local/go/src/testing/testing.go:1117 +0x176
github.com/cilium/cilium/pkg/bpf.runTests(0xc0004e1f00, 0x0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:88 +0x126
github.com/cilium/cilium/pkg/bpf.TestMain(0xc0004e1f00)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:92 +0x2f
main.main()
	_testmain.go:42 +0x135

goroutine 98 [chan receive, 6 minutes]:
gopkg.in/check%2ev1.(*suiteRunner).runTest(...)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:819
gopkg.in/check%2ev1.(*suiteRunner).run(0xc00059fd80, 0x13d0ce0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:624 +0x11d
gopkg.in/check%2ev1.Run(0xcbf880, 0x13d0ce0, 0xc000485708, 0xd0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:92 +0x4d
gopkg.in/check%2ev1.RunAll(0xc00008cf08, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:84 +0x96
gopkg.in/check%2ev1.TestingT(0xc000102600)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:72 +0x395
github.com/cilium/cilium/pkg/bpf.Test(0xc000102600)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:34 +0x2b
testing.tRunner(0xc000102600, 0xd4e2e8)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350

goroutine 37 [syscall, 5 minutes]:
syscall.Syscall(0x141, 0x4, 0xc000092d68, 0x20, 0xc116a0, 0xc0003b4180, 0xc000398770)
	/usr/local/go/src/syscall/asm_linux_amd64.s:18 +0x5
github.com/cilium/cilium/pkg/bpf.GetNextKeyFromPointers(0x5, 0xc000092d68, 0x20, 0xc0002df650, 0x4)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/bpf_linux.go:218 +0x7a
github.com/cilium/cilium/pkg/bpf.(*Map).DumpReliablyWithCallback(0xc0001d6960, 0xc000092f40, 0xc000092f58, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:697 +0x439
github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback.func3(0xc00029c830, 0xc000000100, 0xc0001d6870, 0xc0001d6960, 0xc0000c0840)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:468 +0x315
created by github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:447 +0x467

goroutine 99 [select, 6 minutes]:
gopkg.in/check%2ev1.(*resultTracker)._loopRoutine(0xc0001927e0)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:470 +0xc4
created by gopkg.in/check%2ev1.(*resultTracker).start
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:450 +0x3f

goroutine 84 [semacquire, 6 minutes]:
sync.runtime_Semacquire(0xc00029c838)
	/usr/local/go/src/runtime/sema.go:56 +0x42
sync.(*WaitGroup).Wait(0xc00029c830)
	/usr/local/go/src/sync/waitgroup.go:130 +0x64
github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback(0x13d0ce0, 0xc0001d6870)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:478 +0x478
reflect.Value.call(0xcbf880, 0x13d0ce0, 0x1213, 0xd182fa, 0x4, 0xc00008df30, 0x1, 0x1, 0x13d2be0, 0xc00008ddd0, ...)
	/usr/local/go/src/reflect/value.go:460 +0x5f6
reflect.Value.Call(0xcbf880, 0x13d0ce0, 0x1213, 0xc00008df30, 0x1, 0x1, 0x0, 0xe, 0x8)
	/usr/local/go/src/reflect/value.go:321 +0xb4
gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1(0xc0001d6870)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0x5e4
gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1(0xc00059fd80, 0xc0001d6870, 0xc0003b6c40)
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0xa2
created by gopkg.in/check%2ev1.(*suiteRunner).forkCall
	/home/vagrant/go/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:672 +0x1fb

goroutine 85 [runnable]:
github.com/cilium/cilium/pkg/bpf.(*Map).deleteCacheEntry(0xc0001d6960, 0xe44940, 0xc000640890, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:835 +0x214
github.com/cilium/cilium/pkg/bpf.(*Map).DeleteWithErrno(0xc0001d6960, 0xe44940, 0xc000640890, 0x0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:881 +0x2ca
github.com/cilium/cilium/pkg/bpf.(*Map).Delete(...)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:885
github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback.func2(0xc00029c830, 0xc0000c07e0, 0xc0000c0840, 0xc0001d6960, 0xc0001d6870)
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:437 +0x10e
created by github.com/cilium/cilium/pkg/bpf.(*BPFPrivilegedTestSuite).TestDumpReliablyWithCallback
	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:422 +0x3df
FAIL	github.com/cilium/cilium/pkg/bpf	360.062s
FAIL
Makefile:97: recipe for target 'tests-privileged' failed
make: *** [tests-privileged] Error 1
```
</details>

This time, we have `goroutine 85` trying to delete a map entry (`github.com/cilium/cilium/pkg/bpf.(*Map).deleteCacheEntry()` while `goroutine 37`  is at the `BPF_MAP_GET_NEXT_KEY`  syscall in `github.com/cilium/cilium/pkg/bpf.GetNextKeyFromPointers()` again.

I managed to reproduce this issue several times and the only constant I see is that the `BPF_MAP_GET_NEXT_KEY` syscall hangs. The other goroutines are not always at the same execution point when the test is interrupted by the timeout.
As of note, the tests also succeed quite often on the dev VM.

I'm probably missing something at this point. Could someone else try to reproduce the failure and have a look at it?

--- a/pkg/bpf/bpf_linux.go
+++ b/pkg/bpf/bpf_linux.go
@@ -18,6 +18,7 @@ package bpf
 
 import (
 	"fmt"
+	"io"
 	"os"
 	"path/filepath"
 	"syscall"
@@ -206,7 +207,8 @@ func DeleteElement(fd int, key unsafe.Pointer) error {
 	return nil
 }
 
-// GetNextKeyFromPointers stores, in nextKey, the next key after the key of the map in fd.
+// GetNextKeyFromPointers stores, in nextKey, the next key after the key of the
+// map in fd. When there are no more keys, io.EOF is returned.
 func GetNextKeyFromPointers(fd int, structPtr, sizeOfStruct uintptr) error {
 
 	var duration *spanstat.SpanStat
@@ -221,6 +223,12 @@ func GetNextKeyFromPointers(fd int, structPtr, sizeOfStruct uintptr) error {
 	)
 	if option.Config.MetricsConfig.BPFSyscallDurationEnabled {
 		metrics.BPFSyscallDuration.WithLabelValues(metricOpGetNextKey, metrics.Errno2Outcome(err)).Observe(duration.End(err == 0).Total().Seconds())
+	}
+
+	// BPF_MAP_GET_NEXT_KEY returns ENOENT when all keys have been iterated
+	// translate that to io.EOF to signify there are no next keys
+	if err == syscall.ENOENT {
+		return io.EOF
 	}
 
 	if ret != 0 || err != 0 {
@@ -242,7 +250,8 @@ func GetNextKey(fd int, key, nextKey unsafe.Pointer) error {
 	return GetNextKeyFromPointers(fd, uintptr(unsafe.Pointer(&uba)), unsafe.Sizeof(uba))
 }
 
-// GetFirstKey fetches the first key in the map.
+// GetFirstKey fetches the first key in the map. If there are no keys in the
+// map, io.EOF is returned.
 func GetFirstKey(fd int, nextKey unsafe.Pointer) error {
 	uba := bpfAttrMapOpElem{
 		mapFd: uint32(fd),

--- a/pkg/bpf/map_linux.go
+++ b/pkg/bpf/map_linux.go
@@ -20,6 +20,7 @@ import (
 	"bufio"
 	"context"
 	"fmt"
+	"io"
 	"os"
 	"path"
 	"reflect"
@@ -550,7 +551,10 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
 	}
 
 	if err := GetFirstKey(m.fd, unsafe.Pointer(&nextKey[0])); err != nil {
-		return nil
+		if err == io.EOF {
+			return nil
+		}
+		return err
 	}
 
 	mk := m.MapKey.DeepCopyMapKey()
@@ -590,12 +594,13 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
 
 		copy(key, nextKey)
 
-		err = GetNextKeyFromPointers(m.fd, bpfCurrentKeyPtr, bpfCurrentKeySize)
-		if err != nil {
-			break
+		if err := GetNextKeyFromPointers(m.fd, bpfCurrentKeyPtr, bpfCurrentKeySize); err != nil {
+			if err == io.EOF { // end of map, we're done iterating
+				return nil
+			}
+			return err
 		}
 	}
-	return nil
 }
 
 // DumpWithCallbackIfExists is similar to DumpWithCallback, but returns earlier
@@ -895,7 +900,10 @@ func (m *Map) DeleteAll() error {
 
 	for {
 		if err := GetFirstKey(m.fd, unsafe.Pointer(&nextKey[0])); err != nil {
-			break
+			if err == io.EOF {
+				return nil
+			}
+			return err
 		}
 
 		err := DeleteElement(m.fd, unsafe.Pointer(&nextKey[0]))
@@ -911,8 +919,6 @@ func (m *Map) DeleteAll() error {
 			return err
 		}
 	}
-
-	return nil
 }
 
 // GetNextKey returns the next key in the Map after key.

--- a/pkg/bpf/map_linux.go
+++ b/pkg/bpf/map_linux.go
@@ -19,6 +19,7 @@ package bpf
 import (
 	"bufio"
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"os"
@@ -41,6 +42,10 @@ import (
 	"github.com/sirupsen/logrus"
 	"golang.org/x/sys/unix"
 )
+
+// ErrMaxLookup is returned when the maximum number of map element lookups has
+// been reached.
+var ErrMaxLookup = errors.New("maximum number of lookups reached")
 
 type MapKey interface {
 	fmt.Stringer
@@ -622,8 +627,8 @@ func (m *Map) DumpWithCallbackIfExists(cb DumpCallback) error {
 // additional tracking of the current and recently seen keys, so that if an
 // element is removed from the underlying kernel map during the dump, the dump
 // can continue from a recently seen key rather than restarting from scratch.
-// In addition, it caps the maximum number of map entry iterations by the
-// maximum size of the map.
+// In addition, it caps the maximum number of map entry iterations at 4 times
+// the maximum map size. If this limit is reached, ErrMaxLookup is returned.
 //
 // The caller must provide a callback for handling each entry, and a stats
 // object initialized via a call to NewDumpStats().
@@ -643,12 +648,14 @@ func (m *Map) DumpReliablyWithCallback(cb DumpCallback, stats *DumpStats) error 
 		return err
 	}
 
-	err := GetFirstKey(m.fd, unsafe.Pointer(&currentKey[0]))
-	if err != nil {
-		// Map is empty, nothing to clean up.
+	if err := GetFirstKey(m.fd, unsafe.Pointer(&currentKey[0])); err != nil {
 		stats.Lookup = 1
-		stats.Completed = true
-		return nil
+		if err == io.EOF {
+			// map is empty, nothing to clean up.
+			stats.Completed = true
+			return nil
+		}
+		return err
 	}
 
 	mk := m.MapKey.DeepCopyMapKey()
@@ -671,16 +678,26 @@ func (m *Map) DumpReliablyWithCallback(cb DumpCallback, stats *DumpStats) error 
 	bpfNextKeyPtr := uintptr(unsafe.Pointer(&bpfNextKey))
 	bpfNextKeySize := unsafe.Sizeof(bpfNextKey)
 
-	for stats.Lookup = 1; stats.Lookup <= stats.MaxEntries; stats.Lookup++ {
-		// currentKey was returned by GetNextKey() so we know it existed in the map, but it may have been
-		// deleted by a concurrent map operation. If currentKey is no longer in the map, nextKey will be
-		// the first key in the map again. Use the nextKey only if we still find currentKey in the Lookup()
-		// after the GetNextKey() call, this way we know nextKey is NOT the first key in the map.
-		nextKeyValid := GetNextKeyFromPointers(m.fd, bpfNextKeyPtr, bpfNextKeySize)
+	// maxLookup is an upper bound limit to prevent backtracking forever
+	// when iterating over the map's elements (the map might be concurrently
+	// updated while being iterated)
+	maxLookup := stats.MaxEntries * 4
+
+	// this loop stops when all elements have been iterated
+	// (GetNextKeyFromPointers returns io.EOF) OR, in order to avoid hanging if
+	// the map is continuously updated, when maxLookup has been reached
+	for stats.Lookup = 1; stats.Lookup <= maxLookup; stats.Lookup++ {
+		// currentKey was returned by GetFirstKey()/GetNextKeyFromPointers()
+		// so we know it existed in the map, but it may have been deleted by a
+		// concurrent map operation. If currentKey is no longer in the map,
+		// nextKey will be the first key in the map again. Use the nextKey only
+		// if we still find currentKey in the Lookup() after the
+		// GetNextKeyFromPointers() call, this way we know nextKey is NOT the
+		// first key in the map.
+		nextKeyErr := GetNextKeyFromPointers(m.fd, bpfNextKeyPtr, bpfNextKeySize)
 		err := LookupElementFromPointers(m.fd, bpfCurrentKeyPtr, bpfCurrentKeySize)
 		if err != nil {
 			stats.LookupFailed++
-
 			// Restarting from a invalid key starts the iteration again from the beginning.
 			// If we have a previously found key, try to restart from there instead
 			if prevKeyValid {
@@ -690,8 +707,9 @@ func (m *Map) DumpReliablyWithCallback(cb DumpCallback, stats *DumpStats) error 
 				prevKeyValid = false
 				stats.KeyFallback++
 			} else {
-				// Depending on exactly when currentKey was deleted from the map, nextKey may be the actual
-				// keyelement after the deleted one, or the first element in the map.
+				// Depending on exactly when currentKey was deleted from the
+				// map, nextKey may be the actual key element after the deleted
+				// one, or the first element in the map.
 				copy(currentKey, nextKey)
 				stats.Interrupted++
 			}
@@ -708,18 +726,21 @@ func (m *Map) DumpReliablyWithCallback(cb DumpCallback, stats *DumpStats) error 
 			cb(mk, mv)
 		}
 
-		if nextKeyValid != nil {
-			stats.Completed = true
-			break
+		if nextKeyErr != nil {
+			if nextKeyErr == io.EOF {
+				stats.Completed = true
+				return nil // end of map, we're done iterating
+			}
+			return nextKeyErr
 		}
+
 		// remember the last found key
 		copy(prevKey, currentKey)
 		prevKeyValid = true
 		// continue from the next key
 		copy(currentKey, nextKey)
 	}
-
-	return nil
+	return ErrMaxLookup
 }
 
 // Dump returns the map (type map[string][]string) which contains all

--- a/pkg/bpf/map_linux_test.go
+++ b/pkg/bpf/map_linux_test.go
@@ -19,6 +19,9 @@ package bpf
 import (
 	"fmt"
 	"os"
+	"strconv"
+	"strings"
+	"sync"
 	"testing"
 	"unsafe"
 
@@ -382,6 +385,95 @@ func (s *BPFPrivilegedTestSuite) TestDump(c *C) {
 		"key=105": {"value=205"},
 		"key=106": {"value=206"},
 	})
+}
+
+func (s *BPFPrivilegedTestSuite) TestDumpReliablyWithCallback(c *C) {
+	maxEntries := uint32(256)
+	m := NewMap("cilium_dump_test",
+		MapTypeHash,
+		&TestKey{},
+		int(unsafe.Sizeof(TestKey{})),
+		&TestValue{},
+		int(unsafe.Sizeof(TestValue{})),
+		int(maxEntries),
+		BPF_F_NO_PREALLOC,
+		0,
+		ConvertKeyValue,
+	).WithCache()
+	_, err := m.OpenOrCreate()
+	c.Assert(err, IsNil)
+	defer func() {
+		path, _ := m.Path()
+		os.Remove(path)
+	}()
+	defer m.Close()
+
+	for i := uint32(4); i < maxEntries; i++ {
+		err := m.Update(&TestKey{Key: i}, &TestValue{Value: i + 100})
+		c.Check(err, IsNil) // we want to run the deferred calls
+	}
+	// start a goroutine that continuously updates the map
+	started := make(chan struct{}, 1)
+	done := make(chan struct{}, 1)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		started <- struct{}{}
+		for {
+			for i := uint32(0); i < 4; i++ {
+				if i < 3 {
+					err := m.Update(&TestKey{Key: i}, &TestValue{Value: i + 100})
+					// avoid assert to ensure we call wg.Done
+					c.Check(err, IsNil)
+				}
+				if i > 0 {
+					err := m.Delete(&TestKey{Key: i - 1})
+					// avoid assert to ensure we call wg.Done
+					c.Check(err, IsNil)
+				}
+			}
+			select {
+			case <-done:
+				return
+			default:
+			}
+		}
+	}()
+	<-started // wait until the routine has started to start the actual tests
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		expect := map[string]string{}
+		for i := uint32(4); i < maxEntries; i++ {
+			expect[fmt.Sprintf("key=%d", i)] = fmt.Sprintf("custom-value=%d", i+100)
+		}
+		for i := 0; i < 100; i++ {
+			dump := map[string]string{}
+			customCb := func(key MapKey, value MapValue) {
+				k, err := strconv.ParseUint(strings.TrimPrefix(key.String(), "key="), 10, 32)
+				c.Check(err, IsNil)
+				if uint32(k) >= 4 {
+					dump[key.String()] = "custom-" + value.String()
+				}
+			}
+			ds := NewDumpStats(m)
+			if i == 0 {
+				// artificially trigger MaxLookupError as max lookup is based
+				// on ds.MaxEntries
+				ds.MaxEntries = 1
+			}
+			if err := m.DumpReliablyWithCallback(customCb, ds); err != nil {
+				// avoid Assert to ensure the done signal is sent
+				c.Check(err, Equals, ErrMaxLookup)
+			} else {
+				// avoid Assert to ensure the done signal is sent
+				c.Check(dump, checker.DeepEquals, expect)
+			}
+		}
+		done <- struct{}{}
+	}()
+	wg.Wait()
 }
 
 func (s *BPFPrivilegedTestSuite) TestDeleteAll(c *C) {

